### ISSUE TITLE
OpenAI-compat /v1/chat/completions with tools does not install grammar; model emits invalid tool names

### ISSUE THREAD

**Issue by KrakowiakK** [jundot/omlx#1472]
# OpenAI-compat `/v1/chat/completions` with `tools` does not install grammar; model emits invalid tool names

## Summary

When a client sends an OpenAI-compatible chat completion request that contains `tools` (with or without `tool_choice: "auto"`), oMLX does **not** install a grammar-constrained sampler over the tool-name enum. As a result the model can sample any token sequence as the function name — including shell-command fragments and invented names that are not present in the schema. `tool_call_parser=qwen3_coder_xml` then surfaces those strings verbatim because it is a post-hoc regex parser, not a decoding constraint.

llama.cpp implements this same `qwen3_coder_xml` tool format with a **lazy GBNF grammar** keyed on the request's tool-name enum (`common/chat-parser-xml-toolcall.cpp:168-291`, triggered by `<tool_call>\n<function=`), so the equivalent path on llama.cpp physically cannot emit invalid tool names from the same model.

## Reproducer

- Model: `mlx-community/Qwen3.6-35B-A3B-OptiQ-4bit` (or any Qwen3.6-35B-A3B oQ-quant with `tool_call_parser=qwen3_coder_xml` on the tokenizer)
- oMLX: `0.3.12`, defaults, `enable_thinking=false`
- Request: standard OpenAI-compat `/v1/chat/completions`, with a `tools` array of 4-39 functions (`shell`, `read`, `reason`, `finishTask`, …), `tool_choice: "auto"`, `parallel_tool_calls: true`. Multi-turn coding-agent loop (Exercism JS `space-age` / `bowling`).
- 10 sequential runs of the same task, temperature 0.6, top-p 0.95, top-k 20.

### Observed (oMLX, 10 runs, ~270 tool calls)

```
199 "name":"shell"
 22 "name":"read"
 16 "name":"update_plan"
 12 "name":"reason"
 10 "name":"finishTask"
  6 "name":"cat"                       # not in schema
  2 "name":"ls *.js\n</parameter"      # shell command fragment + leaked closing tag
  2 "name":"ls *.js *.test.* 2"        # shell command fragment
```

3% of tool calls have invalid names. Forty to fifty percent of workflows hit at least one error (dispatch returns `error: unknown tool "<garbled>"`).

### Same model, llama.cpp side-by-side (10 runs, ~230 tool calls)

```
182 "name":"shell"
 40 "name":"reason"
 10 "name":"finishTask"
```

Zero invalid names. Same chat template, same tool schema, same temperature.

## Root cause in oMLX

`server.py:2176` gates grammar compilation strictly on `structured_outputs` / `response_format`:

```python
if request.structured_outputs is not None or response_format:
    compiled_grammar = await _compile_grammar_for_request(...)
```

`_compile_grammar_for_request` (`server.py:2589-2651`) reads `_build_format_element(structured_outputs, response_format)`; with a tools-only request `fmt is None` and the function returns `None` at line 2611. Consequently `scheduler.py:2574` does **not** append a `GrammarConstraintProcessor` to `sampling_params.logits_processors`, and decoding runs unconstrained.

`api/tool_calling.py:378-547` (`qwen3_coder_xml` parser) is reached only after generation. In the regular-content path (`tool_calling.py:630`) the parser does not cross-check the extracted name against `_extract_tool_names(tools)` (that check only runs on think-content at line 658). Whatever string sits between `<function=` and the next delimiter is reported as the tool name.

PR #335 (xgrammar-based constrained decoding, merged 2026-03-29) intentionally scoped grammar installation to `structured_outputs`. PR #995 (open) adds `guided_grammar` aliasing but still does not auto-derive grammar from `tools`. Issue #1258 reports the same architectural gap on the Anthropic `/v1/messages` strict-tool-use path.

## Reference: llama.cpp implementation

For comparison, `common/chat-parser-xml-toolcall.cpp:168-291` (`build_grammar_xml_tool_call`) iterates every tool in the request and emits a GBNF literal alternative per name:

```cpp
// line 203
std::string name = function.at("name");
// lines 259-274
quoted_name = gbnf_format_literal(name);
builder.add_rule(name + "-call",
    gbnf_format_literal(form.tool_start) + " " + quoted_name + " "
    + gbnf_format_literal(form.tool_sep) + " " + next_arg);
// line 277
add_rule("root-tool-call-once", string_join(tool_rules, " | "));
// line 289 — lazy trigger
data.grammar_triggers.push_back({
    COMMON_GRAMMAR_TRIGGER_TYPE_WORD,
    form.scope_start + form.tool_start    // "<tool_call>\n<function="
});
```

`common_chat_params_init_qwen3_coder_xml` (`common/chat.cpp:1850-1880`) sets `data.grammar_lazy = true` and calls the builder above. The grammar is propagated to the sampler via `server-common.cpp:1032-1038` and reinstalled per slot in `server-task.cpp:309-378`. After the trigger string fires, sampling is masked to tokens that match the literal name alternation.

## Proposed fix

Extend the gate at `server.py:2176` so a non-empty `tools` array (or an explicit `tool_choice` other than `"none"`) also triggers grammar compilation. The compiled grammar should be derived from the active `tool_call_parser`'s format. For `qwen3_coder_xml`, this is the same `<tool_call>\n<function=NAME>` shape that llama.cpp emits — i.e. a literal alternation over `function.name` from the request schema, with a lazy trigger keyed on `<function=`.

Behaviour falls back to the current unconstrained path only when the parser has no grammar adapter, so models / parsers without one are unaffected.

## Impact

On Qwen3.6-35B-A3B-class agent workloads we measure:
- oMLX: 40–50 % of workflows hit at least one invalid-name dispatch error (avg 3 % of all tool calls)
- llama.cpp `--jinja` + `tool_call_parser=qwen3_coder_xml`: 0 invalid names across the same workloads

Each invalid emission costs one wasted turn (the agent loop retries with `FORMAT ERROR: unknown tool name`). On long coding-agent runs this materially reduces throughput and inflates `wall` time.

## Environment

- oMLX `0.3.12` (Homebrew, `omlx serve`)
- macOS Sequoia 15.x, Apple M3 Ultra
- Models tested: `Qwen3.6-35B-A3B-oQ5-fp16-mtp` (self-quantized from `Qwen/Qwen3.6-35B-A3B`, `preserve_mtp=true`, `text_only=true`)
- Client: OpenAI-compat over `127.0.0.1:8000/v1/chat/completions`, `tools` array of 4 functions, `tool_choice: auto`, `parallel_tool_calls: true`
- Comparison server: llama.cpp `b9500` with `--jinja` (Qwen3.6 chat template embedded in GGUF)

Happy to share the full JSONL traces (response payloads with the malformed names) and the reproducer scripts on request.

Related: #1258 (Anthropic `/v1/messages` strict tool use ignored — same architectural gap, different endpoint).
